### ISSUE TITLE
PM2 v5.1.0: God Daemon taking up huge amounts of memory

### ISSUE THREAD

**Issue by victor-ono** [Unitech/pm2#5145]
[https://github.com/Unitech/pm2/issues/1126](url)

I'm started experiencing this issue again after updating to 5.1. 

It takes 230MB (50%) of RAM on a 512MB machine

<img width="611" alt="Screen Shot 2021-08-06 at 1 10 24 PM" src="https://user-images.githubusercontent.com/519311/128548181-eee72642-7866-414f-89c1-f426993e35ea.png">

**Comment by yorickshan** [comment]
same issue, any updates？

**Comment by qiulang** [comment]
I have experienced the same problem with 5.2.  The pm2 process takes up about 2.6G memory while the process it controls takes much lower memory ` 15     1 root     S    2645m  34%   1   0% PM2 v5.2.0: God Daemon (/root/.pm2)`. It has lasted for more than 2 days till now.

The following is the output of `pm2 report`,

```
--- PM2 report ----------------------------------------------------------------
Date                 : Fri May 13 2022 14:33:40 GMT+0800 (China Standard Time)
===============================================================================
--- Daemon -------------------------------------------------
pm2d version         : 5.2.0
node version         : 14.18.1
node path            : not found
argv                 : /usr/bin/node,/usr/lib/node_modules/pm2/lib/Daemon.js
argv0                : node
user                 : undefined
uid                  : 0
gid                  : 0
uptime               : 37450min
===============================================================================
--- CLI ----------------------------------------------------
local pm2            : 5.2.0
node version         : 14.18.1
node path            : /usr/bin/pm2
argv                 : /usr/bin/node,/usr/bin/pm2,report
argv0                : node
user                 : undefined
uid                  : 0
gid                  : 0
===============================================================================
--- System info --------------------------------------------
arch                 : x64
platform             : linux
type                 : Linux
cpus                 : Intel(R) Xeon(R) CPU E5-2682 v4 @ 2.50GHz
cpus nb              : 4
freemem              : 227586048
totalmem             : 8201084928
home                 : /root
===============================================================================
--- PM2 list -----------------------------------------------
┌─────┬──────────────────┬─────────────┬─────────┬─────────┬──────────┬────────┬──────┬───────────┬──────────┬──────────┬──────────┬──────────┐
│ id  │ name             │ namespace   │ version │ mode    │ pid      │ uptime │ ↺    │ status    │ cpu      │ mem      │ user     │ watching │
├─────┼──────────────────┼─────────────┼─────────┼─────────┼──────────┼────────┼──────┼───────────┼──────────┼──────────┼──────────┼──────────┤
│ 1   │ ws_redis         │ default     │ 2.0.0   │ fork    │ 9154     │ 24D    │ 5    │ online    │ 0.3%     │ 49.8mb   │ root     │ disabled │
└─────┴──────────────────┴─────────────┴─────────┴─────────┴──────────┴────────┴──────┴───────────┴──────────┴──────────┴──────────┴──────────┘
Module
┌────┬──────────────────────────────┬───────────────┬──────────┬──────────┬──────┬──────────┬──────────┬──────────┐
│ id │ module                       │ version       │ pid      │ status   │ ↺    │ cpu      │ mem      │ user     │
├────┼──────────────────────────────┼───────────────┼──────────┼──────────┼──────┼──────────┼──────────┼──────────┤
│ 0  │ pm2-logrotate                │ 2.7.0         │ 40       │ online   │ 1    │ 0.2%     │ 27.3mb   │ root     │
└────┴──────────────────────────────┴───────────────┴──────────┴──────────┴──────┴──────────┴──────────┴──────────┘
===============================================================================
--- Daemon logs --------------------------------------------
/root/.pm2/pm2.log last 20 lines:
PM2        | 2022-04-19T10:34:07: PM2 log: Stopping app:ws_redis id:1
PM2        | 2022-04-19T10:34:07: PM2 log: App [ws_redis:1] exited with code [0] via signal [SIGINT]
PM2        | 2022-04-19T10:34:07: PM2 log: pid=31052 msg=process killed
PM2        | 2022-04-19T10:34:07: PM2 log: App [ws_redis:1] starting in -fork mode-
PM2        | 2022-04-19T10:34:07: PM2 log: App [ws_redis:1] online
PM2        | 2022-04-19T10:38:26: PM2 log: Stopping app:ws_redis id:1
PM2        | 2022-04-19T10:38:26: PM2 log: App [ws_redis:1] exited with code [0] via signal [SIGINT]
PM2        | 2022-04-19T10:38:26: PM2 log: pid=31208 msg=process killed
PM2        | 2022-04-19T10:38:26: PM2 log: App [ws_redis:1] starting in -fork mode-
PM2        | 2022-04-19T10:38:26: PM2 log: App [ws_redis:1] online
PM2        | 2022-04-19T10:52:09: PM2 log: Stopping app:ws_redis id:1
PM2        | 2022-04-19T10:52:09: PM2 log: App [ws_redis:1] exited with code [0] via signal [SIGINT]
PM2        | 2022-04-19T10:52:09: PM2 log: pid=31497 msg=process killed
PM2        | 2022-04-19T10:52:09: PM2 log: App [ws_redis:1] starting in -fork mode-
PM2        | 2022-04-19T10:52:09: PM2 log: App [ws_redis:1] online
PM2        | 2022-04-19T13:22:44: PM2 log: Stopping app:ws_redis id:1
PM2        | 2022-04-19T13:22:44: PM2 log: App [ws_redis:1] exited with code [0] via signal [SIGINT]
PM2        | 2022-04-19T13:22:44: PM2 log: pid=32384 msg=process killed
PM2        | 2022-04-19T13:22:44: PM2 log: App [ws_redis:1] starting in -fork mode-
PM2        | 2022-04-19T13:22:44: PM2 log: App [ws_redis:1] online

```

**Comment by qiulang** [comment]
The CPU usage is almost always less than 1%, that is why I suspect there is some memory leak somewhere

```
  PID  PPID USER     STAT   VSZ %VSZ CPU %CPU COMMAND
   15     1 root     S    2645m  34%   0   0% PM2 v5.2.0: God Daemon (/root/.pm2)
 9154    15 root     S     312m   4%   2   0% node /home/express/build/index.js
   40    15 root     S     279m   4%   0   0% node /root/.pm2/modules/pm2-logrotate/node_modules/pm2-log
 9175  8997 root     S     261m   3%   2   0% node /usr/bin/pm2 logs
32405 31926 root     T     261m   3%   2   0% node /usr/bin/pm2 logs
31584 31569 root     T     261m   3%   3   0% node /usr/bin/pm2 log 1
  497 31926 root     T     261m   3%   1   0% node /usr/bin/pm2 log 1
  639 31926 root     T     261m   3%   0   0% node /usr/bin/pm2 log 1
  953 31926 root     T     260m   3%   3   0% node /usr/bin/pm2 log 1
```

**Comment by TheAndroidGuy** [comment]
@qiulang This is not problem with PM2, [changing](https://stackoverflow.com/questions/53234410/how-to-use-node-js-with-jemalloc) the memory allocator to `jemalloc` will fix the memory issue.

**Comment by qiulang** [comment]
No I don't think so and https://github.com/nodejs/node/issues/21973 was closed

**Comment by victor-ono** [comment]
I'm still having the same issue.

**Comment by sarojmhrzn** [comment]
I am facing the same issue - when I run the my node application with pm2 it takes huge amount of memory and crashes but when I simply just run "node app.js" then it run without any memory issue and crash.

**Comment by qiulang** [comment]
@sarojmhrzn which node version did you use? Now I use node 16.4 to see if this can fix the problem.

**Comment by EHyang** [comment]
i'm still having the same issue.
i used v5.2.0

**Comment by rodrigograca31** [comment]
same here. pm2 v5.2.0 and seems its not my code because when running with `node file.js --inspect` it stays at about 7mb memory looking at the heap snapshot, while with pm2 it says 47.2mb and keeps going up over time.

**Comment by rodrigograca31** [comment]
also I had 2 similar pieces of code and one didnt have thise memory leak problem.... 
then I noticed that one was using 4.x and updated it to 5.x.... and now in both the memory keeps increasing over time every time I run `pm2 list`

**Comment by smashah** [comment]
The daemon memory leak is insane, on a 256GB RAM VPS, pm2 is managing about 50 puppeteer + chrome isntances and yet pm2 Daemon is taking up 150GB+ of RAM!

HOW can I run the daemon with `--inspect`?? I've fixed memory leaks before but I just need to inspect the process in order to make some headway on this stupid memory leak.

<details>
<summary><b>pm2 report</b></summary>

```
--- PM2 report ----------------------------------------------------------------
Date                 : Fri Dec 16 2022 19:17:07 GMT+0800 (Singapore Standard Time)
===============================================================================
--- Daemon -------------------------------------------------
pm2d version         : 5.2.2
node version         : 18.12.1
node path            : /home/admin/orch/node_modules/.bin/pm2
argv                 : /usr/bin/node,/home/admin/orch/node_modules/pm2/lib/Daemon.js
argv0                : node
user                 : admin
uid                  : 1001
gid                  : 1001
uptime               : 4min
===============================================================================
--- CLI ----------------------------------------------------
local pm2            : 5.2.2
node version         : 18.12.1
node path            : /home/admin/orch/node_modules/.bin/pm2
argv                 : /usr/bin/node,/home/admin/orch/node_modules/.bin/pm2,report
argv0                : node
user                 : admin
uid                  : 1001
gid                  : 1001
===============================================================================
--- System info --------------------------------------------
arch                 : x64
platform             : linux
type                 : Linux
cpus                 : AMD EPYC 7282 16-Core Processor
cpus nb              : 64
freemem              : 242047713280
totalmem             : 270254182400
home                 : /home/admin
===============================================================================
```
</details>

**Comment by smashah** [comment]
tldr: Before the daemon would balloon and take all available memory in the server (256GB RAM) and eventually crash. Now daemon is hovering around 150-300MB of RAM after forcing the daemon to use `jemalloc`.

@rodrigograca31 can you check the above solution and report back if it works for you please. Maybe then I will make a PR to automatically detect and use jemalloc for the Daemon.

**Comment by rodrigograca31** [comment]
I can't anymore... it fixed it self over time if you know what I mean 😂

**Comment by prithvisharma** [comment]
@smashah how do i install `jemalloc` on amazon linux 2? facing the same memory spike issue

**Comment by smashah** [comment]
> @smashah how do i install `jemalloc` on amazon linux 2? facing the same memory spike issue

Try this
`sudo yum install -y jemalloc-devel`

Then use

`npx jemalloc-check` to find the path for jemalloc.

**Comment by victor-ono** [comment]
I stopped using PM2. Instead I'm using Linux native systemd without any memory or other issues.

**Comment by vicatcu** [comment]
I'm running PM2 5.3.0 with Node 18.15.0 on a Raspberry Pi... and top shows

```
812 vic       20   0  188412  71332  33324 R 100.7   3.8   2:56.91 PM2 v5.3.0: God
```

How exactly do I fix this? It's not clear to me from reviewing this thread. I don't know what ecosystem.json is, does it apply to my configuration?

**Comment by vicatcu** [comment]
Thanks I've been using PM2 for a long time and ecosystem.json is new to me... I usually just do `pm2 start run.js --name somename` and then do `pm2 save` and then `pm2 startup` when everything is running the way I want it to. So my followup questions are:
1. Is there a solution for people that use pm2 like me, without an ecosystem.json file, or
2. Is there a 'migration guide' to how to convert a situation like what I described to using ecosystem.json instead?

**Comment by rachitbucha** [comment]
Any update on this issue

**Comment by tintobrazov** [comment]
The problem was the incompatibility of the Node.js and PM2 versions.

Was
node.js 21.6.1
pm2 5.3.1

It became
node.js 18.9.0
pm2 5.3.1

And everything worked as it should

**Comment by zhang771** [comment]
> The problem was the incompatibility of the Node.js and PM2 versions.
> 
> Was node.js 21.6.1 pm2 5.3.1
> 
> It became node.js 18.9.0 pm2 5.3.1
> 
> And everything worked as it should

it works

**Comment by Mikkou** [comment]
pmx didn't help

![image](https://github.com/user-attachments/assets/a588f1b0-ddc3-4953-9256-d7466dc3a9b7)

config
module.exports = {
  apps: [
    {
      name: 'app-3160',
      script: 'dist/src/main.js',
      exec_mode: 'cluster',
      instances: 2,
      watch: false,
      max_memory_restart: '4G',
      listen_timeout: 22000,
      kill_timeout: 8000,
      restart_delay: 4000,
      time: true,
      cwd: '.',
      wait_ready: true,
      instance_var: 'APP_INSTANCE_SEQ',
      pmx: false,
      env: {
        ENV: 'prod',
        NODE_ENV: 'production',
        NODE_OPTIONS: '--max-old-space-size=4096',
      },
    },
  ],
};
